examples: FetchBlock: enhance this example to be able to connect dns discovered peers so that you don't need to install bitcoind locally

Bo-Ye opened the pull request:
In the examples, FetchBlock example only can retrieve a block from localhost, which is inconvenient and especially confusing for the newbies because localhost node needs to install bitcoind, etc.

This commit provides extra feature to download blocks from DNS seeds generating peers, by using --localhost flag on the command line.

Please review and leave comments, thanks.

schildbach commented:
Thanks for your contribution! Generally I'm happy about this enhancement.

I'm not so happy about the introduced dependency to commons-cli. We already use [jopt-simple](https://pholser.github.io/jopt-simple/) for command line parsing, is there any reason to switch?

Bo-Ye commented:
@schildbach Sorry for late reply, thanks for the comments, I didn't realise jopt-simple is available in the first place, I've corrected my PR as per your suggestion, please review again, many thanks.

The PR is rebased and squashed.

schildbach commented on examples/src/main/java/org/bitcoinj/examples/FetchBlock.java:
```diff
@@ -17,34 +17,71 @@
 
 /**
- * Downloads the block given a block hash from the localhost node and prints it out.
+ * Downloads the block given a block hash from the remote/localhost node and prints it out.
```
Paragraphs are missing closing tags. Also, the first paragraph needs markup too.

Bo-Ye replied:
good point, corrected.

schildbach commented on examples/src/main/java/org/bitcoinj/examples/FetchBlock.java:
```diff
@@ -17,34 +17,71 @@
+ * Downloads the block given a block hash from the remote/localhost node and prints it out.
+ * <p>
+ * Downloading from localhost node needs to install bitcoin https://bitcoin.org/en/download and run command: bitcoind -testnet -daemon
+ * After bitcoind up and running, use command: org.bitcoinj.examples.FetchBlock --localhost <blockHash>
```
I suspect `<blockHash>` needs to be escaped.

Bo-Ye replied:
good point, corrected.

schildbach commented on examples/src/main/java/org/bitcoinj/examples/FetchBlock.java:
```diff
@@ -17,34 +17,71 @@
+                return;
+            }
+            nonOpts = opts.nonOptionArguments();
+            if(nonOpts.size() != 1){
```
Space after `if`.

Bo-Ye replied:
good point, corrected.

schildbach commented on examples/src/main/java/org/bitcoinj/examples/FetchBlock.java:
```diff
@@ -17,34 +17,71 @@
+            if(nonOpts.size() != 1){
+                throw new IllegalArgumentException("Incorrect number of block hash, please provide only one block hash.");
+            }
+        }catch (OptionException | IllegalArgumentException e) {
```
Space before `catch`.

Bo-Ye replied:
good point, corrected.

schildbach commented:
I'm currently not sure why the test has failed, but it can't be caused by your changes.

Bo-Ye commented:
Thanks for the comments, I've corrected them in the new commit which is rebased and squashed.

But this time the CI build failed again, I am not able to reproduce in my local, and it seems the failed tests are random, the failure this time is not same as failure last time.

Let me know any discoveries for the failing CI build, it's weird though. Thanks.

Bo-Ye commented:
Finally, I got a green build after rebasing my pull request!

schildbach commented:
Your test failures will likely be fixed with this PR: #1412. I'll try to merge it soon if no one objects.

I'll review your PR shortly.

schildbach commented:
Merged.

Bo-Ye commented:
cook, thanks.
